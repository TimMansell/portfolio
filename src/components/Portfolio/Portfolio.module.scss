@import 'utilities/pxtorem';

.portfolio {
  display: grid;
  grid-gap: pxtorem(24);

  @media (min-width: $bp-lg) {
    grid-template-columns: repeat(2, 1fr);
  }

  @media (min-width: $bp-xl) {
    grid-template-columns: repeat(3, 1fr);
  }
}
