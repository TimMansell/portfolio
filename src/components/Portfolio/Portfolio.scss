@import 'utilities/breakpoints';

.portfolio {
  display: grid;
  grid-gap: 24px;

  @include media-breakpoint-up(sm) {
    grid-template-columns: repeat(2, 1fr);
  }

  @include media-breakpoint-up(lg) {
    grid-template-columns: repeat(3, 1fr);
  }
}
