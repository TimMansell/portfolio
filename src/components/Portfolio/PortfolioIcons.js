import React from 'react';

import './PortfolioIcons.scss';

export const PortfolioIcons = () => {
  return <div className="portfolio-icons">
    <div className="portfolio-icons__icon" />
    <div className="portfolio-icons__icon" />
    <div className="portfolio-icons__icon" />
  </div>;
};

export default PortfolioIcons;
