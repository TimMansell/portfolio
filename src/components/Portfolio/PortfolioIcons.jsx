import React from 'react';
import PropTypes from 'prop-types';

import styles from './PortfolioIcons.module.scss';

export const PortfolioIcons = ({ title }) => {
  return (
    <div className={styles.bar}>
      <div className={styles.icons}>
        <div className={styles.icon} />
        <div className={styles.icon} />
        <div className={styles.icon} />
      </div>
      <div className={styles.tab}>{title}</div>
    </div>
  );
};

export default PortfolioIcons;

PortfolioIcons.propTypes = {
  title: PropTypes.string.isRequired,
};
