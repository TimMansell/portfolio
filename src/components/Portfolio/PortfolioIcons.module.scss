@import 'utilities/pxtorem';

:root {
  --icon-background: hsl(240, 3%, 89%);
  --icon-tab: hsl(240, 3%, 80%);
  --icon-colour-primary: hsl(3, 100%, 67%);
  --icon-colour-primary-shadow: hsl(3, 100%, 42%);
  --icon-colour-secondary: hsl(41, 100%, 59%);
  --icon-colour-secondary-shadow: hsl(41, 100%, 34%);
  --icon-colour-tertiary: hsl(130, 67%, 47%);
  --icon-colour-tertiary-shadow: hsl(130, 67%, 22%);
  --tab-colour: hsl(0, 0%, 35%);
}

.bar {
  display: flex;
  background: var(--icon-background);
  border-bottom: 1px solid var(--icon-tab);
  overflow: hidden;
  height: var(--vr-md);
}

.icons {
  display: flex;
  padding-left: var(--vr-xs);
  padding-right: var(--vr-xs);
  align-items: center;
}

.icon {
  width: pxtorem(10);
  height: pxtorem(10);
  border-radius: 50%;
  margin-right: var(--vr-xs);

  &:first-of-type {
    background: var(--icon-colour-primary);
    box-shadow: inset 0 0 1px var(--icon-colour-primary-shadow);
  }

  &:nth-of-type(2) {
    background: var(--icon-colour-secondary);
    box-shadow: inset 0 0 1px var(--icon-colour-secondary-shadow);
  }

  &:last-of-type {
    background: var(--icon-colour-tertiary);
    box-shadow: inset 0 0 1px var(--icon-colour-tertiary-shadow);
  }
}

.tab {
  align-self: flex-end;
  position: relative;
  background: var(--icon-background);
  border-top-left-radius: 5px;
  border-top-right-radius: 5px;
  padding-left: var(--vr-xs);
  padding-right: var(--vr-xs);
  font-size: pxtorem(12);
  border: 1px solid var(--icon-tab);
  border-bottom: 0;
  line-height: 2.2;
  color: var(--tab-colour);
  backface-visibility: hidden;

  &::before,
  &::after {
    content: '';
    position: absolute;
    bottom: 0;
    height: pxtorem(16);
    width: pxtorem(16);
  }

  &::before {
    left: pxtorem(-16);
    border-right: 1px solid var(--icon-tab);
    border-bottom-right-radius: 8px;
    box-shadow: 5px 5px 0 var(--icon-background);
  }

  &::after {
    right: pxtorem(-16);
    border-bottom-left-radius: 8px;
    border-left: 1px solid var(--icon-tab);
    box-shadow: -5px 5px 0 var(--icon-background);
  }
}
