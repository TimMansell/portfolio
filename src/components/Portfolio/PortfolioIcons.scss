$icon-colour-primary: #ff5f58;
$icon-colour-secondary: #ffbd2e;
$icon-colour-tertiary: #28ca42;

.portfolio-icons {
	display: flex;
	padding: 10px;

	&__icon {
		width: 10px;
		height: 10px;
		border-radius: 50%;
		margin-right: 8px;

		&:first-of-type {
			background: $icon-colour-primary;
			box-shadow: inset 0 0 1px darken($icon-colour-primary, 25%);
		}

		&:nth-of-type(2) {
			background: $icon-colour-secondary;
			box-shadow: inset 0 0 1px darken($icon-colour-secondary, 25%);
		}

		&:last-of-type {
			background: $icon-colour-tertiary;
			box-shadow: inset 0 0 1px darken($icon-colour-tertiary, 25%);
		}
	}
}