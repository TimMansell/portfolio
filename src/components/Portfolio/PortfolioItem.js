import React, { useState, useEffect, useRef } from 'react';
import PropTypes from 'prop-types';
import debounce from 'lodash.debounce';

import Button from 'components/Button';
import PortfolioIcons from './PortfolioIcons';
import Picture from '../Picture';
import Labels from '../Labels';

import useInViewport from 'hooks/useInViewport';

import styles from './PortfolioItem.module.scss';

export const PortfolioItem = ({
  src,
  title,
  year,
  description,
  tech,
  url,
  source,
}) => {
  const [cubeStyles, setCubeStyles] = useState({});
  const cubeElement = useRef(null);
  const isInViewport = useInViewport(cubeElement, 300);

  const { name, types, fallback } = src;
  const srcs = types.map((type) => ({
    type,
    src: require(`./img/${name}.${type}`),
  }));
  const defaultImg = require(`./img/${name}.${fallback}`);

  const getCubeHeight = debounce(() => {
    const { clientHeight } = cubeElement.current;

    setCubeStyles({
      transformOrigin: `50% 50% -${clientHeight / 2}px`,
    });
  });

  // Check height of image as intially it is lazy loaded.
  useEffect(() => {
    window.requestAnimationFrame(getCubeHeight);

    return () => window.cancelAnimationFrame(getCubeHeight);
  });

  return (
    <div className={styles.portfolioItem}>
      <div className={styles.cube} style={cubeStyles} ref={cubeElement}>
        <div className={styles.browser}>
          <PortfolioIcons title={title} />
          {isInViewport && (
            <Picture srcs={srcs} name={title} defaultImg={defaultImg} />
          )}
        </div>

        <div className={styles.info}>
          <h3 className={styles.title}>{title}</h3>
          <p className={styles.description}>{description}</p>
          <div className={styles.buttons}>
            <Button
              href={url}
              target="_blank"
              title="Visit website"
              data-test="portfolio-website-btn"
            >
              Visit website
            </Button>
            {source && (
              <Button
                href={source}
                target="_blank"
                title="View Souce"
                data-test="portfolio-source-btn"
              >
                View source
              </Button>
            )}
          </div>
        </div>
      </div>
      <div className={styles.tech} data-test="portfolio-tech">
        <Labels items={[year]} type="secondary" />
      </div>
      <Labels items={tech} type="primary" />
    </div>
  );
};

export default PortfolioItem;

PortfolioItem.propTypes = {
  src: PropTypes.object.isRequired,
  title: PropTypes.string.isRequired,
  year: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
  tech: PropTypes.array.isRequired,
  url: PropTypes.string.isRequired,
  source: PropTypes.string,
};
