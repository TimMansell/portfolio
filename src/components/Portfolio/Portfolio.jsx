import Header from 'components/Header';
import PortfolioItem from './PortfolioItem';

import styles from './Portfolio.module.scss';
import portfolioJson from './json/portfolio.json';

export const Portfolio = () => {
  const listItems = portfolioJson.map((item, i) => (
    <PortfolioItem {...item} key={i} />
  ));

  return (
    <>
      <Header
        title="Portfolio"
        text="Here's a sample of my work. Hover over image for more information"
        type="tertiary"
      />

      <div className={styles.portfolio} data-e2e="portfolio-items">
        {listItems}
      </div>
    </>
  );
};

export default Portfolio;
