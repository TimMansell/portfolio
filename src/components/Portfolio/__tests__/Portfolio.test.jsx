import { it, describe, expect } from 'vitest';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

import Portfolio from '../Portfolio';

vi.mock('../json/portfolio.json', () => ({
  default: [
    {
      title: 'title',
      years: ['2020'],
      types: ['types'],
      description: 'description',
      tech: ['tech 1', 'tech 2'],
      url: 'url',
      source: 'source',
      img: 'test',
    },
    {
      title: 'title',
      years: ['2020', '2021'],
      types: ['types'],
      description: 'description',
      tech: ['tech 1', 'tech 2', 'tech 3'],
      url: 'url',
      img: 'test',
    },
  ],
}));

describe('Portfolio', () => {
  it('should render my component', () => {
    // eslint-disable-next-line
    const wrapper = shallow(<Portfolio />);
  });

  it('should match snapshot', () => {
    const snapshot = renderer.create(<Portfolio />).toJSON();

    expect(snapshot).toMatchSnapshot();
  });
});
