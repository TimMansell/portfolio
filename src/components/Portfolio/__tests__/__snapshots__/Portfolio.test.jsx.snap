// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Portfolio > should match snapshot 1`] = `
[
  <div
    className="heading headingTertiary"
  >
    <h2
      aria-label="Portfolio"
      className="title"
      data-test="heading-title"
    >
      Portfolio
    </h2>
    <p
      className="description"
      data-test="heading-description"
    >
      Here's a sample of my work. Hover over image for more information
    </p>
  </div>,
  <div
    className="portfolio"
    data-e2e="portfolio-items"
  >
    <div
      className="portfolioItem"
    >
      <div
        className="cube"
        style={{}}
      >
        <div
          className="browser"
        >
          <div
            className="bar"
          >
            <div
              className="icons"
            >
              <div
                className="icon"
              />
              <div
                className="icon"
              />
              <div
                className="icon"
              />
            </div>
            <div
              className="tab"
            >
              title
            </div>
          </div>
          <picture
            data-e2e="picture"
            title="title portfolio item"
          >
            <source
              data-test="picture-source-0"
              sizes="(min-width: 1200px) 33vw, (min-width: 768px) 50vw, 100vw"
              srcSet="file:///H:/projects/portfolio/src/components/Portfolio/img/webp/test-480.webp 480w, file:///H:/projects/portfolio/src/components/Portfolio/img/webp/test-640.webp 640w, file:///H:/projects/portfolio/src/components/Portfolio/img/webp/test-768.webp 768w, file:///H:/projects/portfolio/src/components/Portfolio/img/webp/test-1024.webp 1024w, file:///H:/projects/portfolio/src/components/Portfolio/img/webp/test-1280.webp 1280w"
              type="image/webp"
            />
            <source
              data-test="picture-source-1"
              sizes="(min-width: 1200px) 33vw, (min-width: 768px) 50vw, 100vw"
              srcSet="file:///H:/projects/portfolio/src/components/Portfolio/img/jpg/test-480.jpg 480w, file:///H:/projects/portfolio/src/components/Portfolio/img/jpg/test-640.jpg 640w, file:///H:/projects/portfolio/src/components/Portfolio/img/jpg/test-768.jpg 768w, file:///H:/projects/portfolio/src/components/Portfolio/img/jpg/test-1024.jpg 1024w, file:///H:/projects/portfolio/src/components/Portfolio/img/jpg/test-1280.jpg 1280w"
              type="image/jpg"
            />
            <img
              alt="title portfolio item"
              className="img"
              data-e2e="picture-img"
              data-test="picture-img"
              loading="lazy"
              src="file:///H:/projects/portfolio/src/components/Portfolio/img/jpg/test-1280.jpg"
            />
          </picture>
        </div>
        <div
          className="info"
        >
          <h3
            className="title"
          >
            title
          </h3>
          <p
            className="description"
          >
            description
          </p>
          <div
            className="buttons"
          >
            <a
              className="btn"
              href="url"
              rel="noopener noreferrer"
              target="_blank"
              title="Visit website"
            >
              Visit website
            </a>
            <a
              className="btn"
              href="source"
              rel="noopener noreferrer"
              target="_blank"
              title="View Souce"
            >
              View source
            </a>
          </div>
        </div>
      </div>
      <div
        className="labels"
      >
        <ul
          className="labels"
          data-test="labels"
        >
          <li
            className="label isSecondary"
          >
            2020
          </li>
          <li
            className="label isSecondary"
          >
            types
          </li>
          <li
            className="label isPrimary"
          >
            tech 1
          </li>
          <li
            className="label isPrimary"
          >
            tech 2
          </li>
        </ul>
      </div>
    </div>
    <div
      className="portfolioItem"
    >
      <div
        className="cube"
        style={{}}
      >
        <div
          className="browser"
        >
          <div
            className="bar"
          >
            <div
              className="icons"
            >
              <div
                className="icon"
              />
              <div
                className="icon"
              />
              <div
                className="icon"
              />
            </div>
            <div
              className="tab"
            >
              title
            </div>
          </div>
          <picture
            data-e2e="picture"
            title="title portfolio item"
          >
            <source
              data-test="picture-source-0"
              sizes="(min-width: 1200px) 33vw, (min-width: 768px) 50vw, 100vw"
              srcSet="file:///H:/projects/portfolio/src/components/Portfolio/img/webp/test-480.webp 480w, file:///H:/projects/portfolio/src/components/Portfolio/img/webp/test-640.webp 640w, file:///H:/projects/portfolio/src/components/Portfolio/img/webp/test-768.webp 768w, file:///H:/projects/portfolio/src/components/Portfolio/img/webp/test-1024.webp 1024w, file:///H:/projects/portfolio/src/components/Portfolio/img/webp/test-1280.webp 1280w"
              type="image/webp"
            />
            <source
              data-test="picture-source-1"
              sizes="(min-width: 1200px) 33vw, (min-width: 768px) 50vw, 100vw"
              srcSet="file:///H:/projects/portfolio/src/components/Portfolio/img/jpg/test-480.jpg 480w, file:///H:/projects/portfolio/src/components/Portfolio/img/jpg/test-640.jpg 640w, file:///H:/projects/portfolio/src/components/Portfolio/img/jpg/test-768.jpg 768w, file:///H:/projects/portfolio/src/components/Portfolio/img/jpg/test-1024.jpg 1024w, file:///H:/projects/portfolio/src/components/Portfolio/img/jpg/test-1280.jpg 1280w"
              type="image/jpg"
            />
            <img
              alt="title portfolio item"
              className="img"
              data-e2e="picture-img"
              data-test="picture-img"
              loading="lazy"
              src="file:///H:/projects/portfolio/src/components/Portfolio/img/jpg/test-1280.jpg"
            />
          </picture>
        </div>
        <div
          className="info"
        >
          <h3
            className="title"
          >
            title
          </h3>
          <p
            className="description"
          >
            description
          </p>
          <div
            className="buttons"
          >
            <a
              className="btn"
              href="url"
              rel="noopener noreferrer"
              target="_blank"
              title="Visit website"
            >
              Visit website
            </a>
          </div>
        </div>
      </div>
      <div
        className="labels"
      >
        <ul
          className="labels"
          data-test="labels"
        >
          <li
            className="label isSecondary"
          >
            2020
          </li>
          <li
            className="label isSecondary"
          >
            2021
          </li>
          <li
            className="label isSecondary"
          >
            types
          </li>
          <li
            className="label isPrimary"
          >
            tech 1
          </li>
          <li
            className="label isPrimary"
          >
            tech 2
          </li>
          <li
            className="label isPrimary"
          >
            tech 3
          </li>
        </ul>
      </div>
    </div>
  </div>,
]
`;
